sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/KafkaBundleMetadataContextProducerIntegrationTest.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junitpioneer.jupiter.SetSystemProperty;
+import org.sdase.commons.server.dropwizard.metadata.DetachedMetadataContext;
+import org.sdase.commons.server.dropwizard.metadata.MetadataContext;
+import org.sdase.commons.server.kafka.builder.ProducerRegistration;
+import org.sdase.commons.server.kafka.dropwizard.KafkaTestApplication;
+import org.sdase.commons.server.kafka.dropwizard.KafkaTestConfiguration;
+import org.sdase.commons.server.kafka.producer.MessageProducer;
+
+@SetSystemProperty(key = "METADATA_FIELDS", value = "tenant-id")
+class KafkaBundleMetadataContextProducerIntegrationTest {
+
+  @RegisterExtension
+  @Order(0)
+  private static final SharedKafkaTestResource KAFKA = new SharedKafkaTestResource();
+
+  @RegisterExtension
+  @Order(1)
+  private static final DropwizardAppExtension<KafkaTestConfiguration> DROPWIZARD_APP_EXTENSION =
+      new DropwizardAppExtension<>(
+          KafkaTestApplication.class,
+          resourceFilePath("test-config-default.yml"),
+          config("kafka.brokers", KAFKA::getKafkaConnectString),
+
+          // performance improvements in the tests
+          config("kafka.config.heartbeat\\.interval\\.ms", "250"),
+          config("kafka.adminConfig.adminClientRequestTimeoutMs", "30000"));
+
+  private static final String TOPIC = "metadata-producer";
+
+  private int messagesBefore = 0;
+
+  private MessageProducer<String, String> kafkaProducer;
+
+  @BeforeAll
+  static void createTopic() {
+    KAFKA.getKafkaTestUtils().createTopic(TOPIC, 1, (short) 1);
+  }
+
+  @BeforeEach
+  void before() {
+    messagesBefore = KAFKA.getKafkaTestUtils().consumeAllRecordsFromTopic(TOPIC).size();
+
+    KafkaTestApplication app = DROPWIZARD_APP_EXTENSION.getApplication();
+    KafkaBundle<KafkaTestConfiguration> kafkaBundle = app.kafkaBundle();
+
+    ProducerRegistration<String, String> producerRegistration =
+        ProducerRegistration.builder()
+            .forTopic(TOPIC)
+            .withDefaultProducer()
+            .withKeySerializer(new StringSerializer())
+            .withValueSerializer(new StringSerializer())
+            .build();
+    kafkaProducer = kafkaBundle.registerProducer(producerRegistration);
+  }
+
+  @BeforeEach
+  @AfterEach
+  void clearMetadata() {
+    MetadataContext.createContext(new DetachedMetadataContext());
+  }
+
+  @Test
+  void shouldSendMetadataContext() {
+    DetachedMetadataContext given = new DetachedMetadataContext();
+    given.put("tenant-id", List.of("t-1"));
```

Comment:
Can we provide a fluent syntax to our clients?
ie. `new DetachedMetadataContext().put("foo", "bar")`

This example could make use of a varargs method. Would make it easier to use.

Reply:
We extend a `LinkedHashMap` here. And that's the API of `Map`. Imho, we loose a lot of other benefits if we implement all that on our own.